packages/base-cli/src/index.ts

```diff
@@ -0,0 +1,212 @@
+import fs from 'fs';
+import path from 'path';
+import { parseArguments, SignalValue } from './args';
+import postcss from 'postcss';
+import type { PluginCreator } from 'postcss';
+
+export * from './help';
+
+type PluginCreatorOptions = Record<string, unknown> | null;
+
+export function cli(plugin: PluginCreator<PluginCreatorOptions>, allowedPluginOpts: Array<string>, helpLogger: () => void) {
+	// get process and plugin options from the command line
+	const argo = parseArguments(process.argv.slice(2), allowedPluginOpts, helpLogger);
+	if (argo === SignalValue.InvalidArguments) {
+		process.exit(1);
+	}
+
+	if (argo.stdin && argo.stdout) {
+		return getStdin().then((css) => {
+			if (!css) {
+				helpLogger();
+				process.exit(1);
+			}
+
+			const result = postcss([plugin(argo.pluginOptions)]).process(css, {
+				from: 'stdin',
+				to: 'stdout',
+				map: argo.inlineMap ? { inline: true } : false,
+			});
+
+			return result.css;
+		}).then((result) => {
+			process.stdout.write(result + (argo.inlineMap ? '\n' : ''));
+
+			process.exit(0);
+		}).catch((error) => {
+			console.error(argo.debug ? error : error.message);
+
+			process.exit(1);
+		});
+	}
+
+	if (argo.stdin) {
+		let output = argo.output;
+		if (!output && argo.outputDir) {
+			output = path.join(argo.outputDir, 'output.css');
+		}
+
+		return getStdin().then((css) => {
+			if (!css) {
+				helpLogger();
+				process.exit(1);
+			}
```

Comment:
All of this logic feels rather similar to the above, should we extract or is this a ca see of keeping this as is to see where it could deviate?

Reply:
Just read your comment below, disregard this one. However I feel like some of the things could be extracted into their own functions?

Reply:
I split up the code from `index.ts` into 4 separate files with one function each.
This does not remove much of the duplication.
But each case is different in subtle ways.

Just splitting it up and using `async`/`await` makes each code path a bit more readable (imho).